README.md
# dramatiq-azure
[![CI](https://github.com/bidossessi/dramatiq-azure/actions/workflows/ci.yml/badge.svg)](https://github.com/bidossessi/dramatiq-azure/actions/workflows/ci.yml)
[![Pypi](https://github.com/bidossessi/dramatiq-azure/actions/workflows/python-publish.yml/badge.svg)](https://github.com/bidossessi/dramatiq-azure/actions/workflows/python-publish.yml)
[![codecov](https://codecov.io/gh/bidossessi/dramatiq-azure/branch/main/graph/badge.svg?token=6LLEDAM3SG)](https://codecov.io/gh/bidossessi/dramatiq-azure)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://makeapullrequest.com)


A [Dramatiq](https://dramatiq.io) broker that can be used with [Microsoft Azure](https://azure.microsoft.com/en-us/) queue services.

Heavily inspired by [Dramatiq SQS](https://github.com/Bogdanp/dramatiq_sqs), this package currently implements a broker for [Azure Storage Queue](https://docs.microsoft.com/en-us/azure/storage/queues/).
An implementation for [Azure Service Bus](https://docs.microsoft.com/en-us/azure/service-bus-messaging/) is planned... eventually.


## Installation

```shell
    pip install dramatiq-azure
    pip install dramatiq-azure[identity]  # for passwordless authentication
```
## Usage

### ASQBroker

```python
import dramatiq

from dramatiq.middleware import AgeLimit, TimeLimit, Callbacks, Pipelines, Prometheus, Retries
from dramatiq_azure import ASQBroker


broker = ASQBroker(
    dead_letter=True,
    middleware=[
        Prometheus(),
        AgeLimit(),
        TimeLimit(),
        Callbacks(),
        Pipelines(),
        Retries(min_backoff=1000, max_backoff=900000, max_retries=96),
    ],
)
dramatiq.set_broker(broker)
```

### Authentication

The following authentication methods are supported by the broker:
1. Connection string based: `AZURE_STORAGE_CONNECTION_STR` environment variable must be set.
If this variable is not set, passwordless authentication will be used.
Creating a connection string for your Azure account is documented [here](https://docs.microsoft.com/en-us/azure/storage/common/storage-configure-connection-string).
2. [Passwordless](https://learn.microsoft.com/en-us/azure/developer/python/sdk/authentication-overview#recommended-app-authentication-approach) (token-based) authentication **(Recommended)**: `AZURE_STORAGE_ACCOUNT_NAME` environment variable must be set.

The list of other mandatory variables depends on where the app is being run.
More information can be found [here](https://learn.microsoft.com/en-us/azure/storage/queues/storage-quickstart-queues-python?tabs=passwordless%2Croles-azure-portal%2Cenvironment-variable-windows%2Csign-in-azure-cli#authenticate-to-azure).

### Environment variables

The following environment variables can be used to configure the broker:
- `AZURE_STORAGE_CONNECTION_STR`: Azure Storage connection string;
- `AZURE_STORAGE_ACCOUNT_NAME`/`AZURE_ACCOUNT_NAME`: Azure Storage account name;
- `AZURE_ENDPOINT_SUFFIX`: Azure Storage endpoint suffix;
- `AZURE_SSL`: Whether to use SSL for the connection;
- `AZURE_QUEUE_ACCOUNT_URL`: Azure Storage account URL;
- `DRAMATIQ_ASQ_MIN_TIMEOUT`: The minimum time to wait between polls in second.

## Tests

Tests require a running [Azurite](https://github.com/Azure/Azurite) instance. You can easily launch `azurite` through [Docker](https://www.docker.com/).

```shell
docker run -p 10000:10000 -p 10001:10001 -p 10002:10002 mcr.microsoft.com/azure-storage/azurite
```

Run the test suite

```shell
pytest
```

## Contributions

Found an itch you know how to scratch? PR welcome (just remember to read the
[contribution guide](CONTRIBUTING.md)) !


CHANGELOG.md
## v0.3.1 (2023-09-01)

### Refactor

- improve environment variable options for passwordless authentication

## v0.3.0 (2023-08-21)

### Feat

- Add DefaultAzureCredential

## v0.2.0 (2023-07-16)

### Feat

- allow overriding the default consumer

## v0.1.6 (2023-07-14)

### Refactor

- fix style (#12)

## v0.1.5 (2022-03-02)

## v0.1.4 (2022-02-11)

## v0.1.3 (2022-02-10)

## v0.1.2 (2022-02-10)

## v0.1.1 (2022-02-09)


tests/conftest.py
import random
import string

import dramatiq
import pytest
from dramatiq.middleware import (
    AgeLimit,
    Callbacks,
    Pipelines,
    Retries,
    TimeLimit,
)

from dramatiq_azure import asq


@pytest.fixture
def asq_broker():
    broker = asq.ASQBroker(
        dead_letter=True,
        middleware=[
            AgeLimit(),
            TimeLimit(),
            Callbacks(),
            Pipelines(),
            Retries(min_backoff=1000, max_backoff=900000, max_retries=96),
        ],
    )
    dramatiq.set_broker(broker)
    yield broker
    for queue_name in broker.queues:
        client = asq._get_client(queue_name)
        client.delete_queue()
        if broker.dead_letter:
            dlq_client = asq._get_dlq_client(queue_name)
            dlq_client.delete_queue()


@pytest.fixture
def queue_name():
    letters = string.ascii_lowercase
    result_str = "".join(random.choice(letters) for i in range(7))
    return f"queue{result_str}"


@pytest.fixture
def worker(asq_broker):
    worker = dramatiq.Worker(asq_broker, worker_threads=1)
    worker.start()
    yield worker
    worker.stop()


tests/test_asq.py
import time

import dramatiq
import pytest
from azure.storage.queue import QueueClient

from dramatiq_azure import asq


def test_can_enqueue_and_process_messages(asq_broker, worker, queue_name):
    # Given an actor that stores incoming messages in a database
    db = []

    @dramatiq.actor(queue_name=queue_name)
    def do_work(x):
        db.append(x)

    # When I send that actor a message
    do_work.send(1)

    # And wait for it to be processed
    time.sleep(5)

    # Then the db should contain that message
    assert db == [1]


def test_limits_prefetch_if_consumer_queue_is_full(
    asq_broker, worker, queue_name
):
    # Given that I have an actor that stores incoming messages in a database
    db = []

    # Set the worker prefetch limit to 1
    worker.queue_prefetch = 1

    # Add delay to actor logic to simulate processing time
    @dramatiq.actor(queue_name=queue_name)
    def do_work(x):
        db.append(x)
        time.sleep(10)

    # When I send that actor messages, it'll only prefetch and process a single message
    do_work.send(1)
    do_work.send(2)

    # Wait for message to be processed
    time.sleep(5)

    # Then the db should contain only that message, while it sleeps
    assert len(db) == 1


def test_can_enqueue_delayed_messages(asq_broker, worker, queue_name):
    # Given an actor that stores incoming messages in a database
    db = []

    @dramatiq.actor(queue_name=queue_name)
    def do_work(x):
        db.append(x)

    # When I send that actor a delayed message
    start_time = time.time()
    do_work.send_with_options(args=(1,), delay=5000)

    # And poll the database for a result each second
    for _ in range(60):
        if db:
            break

        time.sleep(1)

    # Then the db should contain that message
    assert db == [1]

    # And an appropriate amount of time should have passed
    delta = time.time() - start_time
    assert delta >= 5


def test_cant_delay_messages_for_longer_than_7_days(asq_broker, queue_name):
    # Given that I have an actor
    @dramatiq.actor(queue_name=queue_name)
    def do_work():
        pass

    # When I attempt to send that actor a message farther than 7 days into the future
    # Then I should get back a RuntimeError
    with pytest.raises(RuntimeError):
        do_work.send_with_options(delay=7 * 25 * 60 * 60 * 1000)


def test_cant_enqueue_messages_that_are_too_large(asq_broker, queue_name):
    # Given that I have an actor
    @dramatiq.actor(queue_name=queue_name)
    def do_work(s):
        pass

    # When I attempt to send that actor a message that's too large
    # Then a RuntimeError should be raised
    with pytest.raises(RuntimeError):
        do_work.send("a" * 64 * 1024)


def test_can_requeue_consumed_messages(asq_broker, queue_name):
    # Given an actor
    @dramatiq.actor(queue_name=queue_name)
    def do_work():
        pass

    # When I send that actor a message
    do_work.send()

    # And consume the message off the queue
    consumer = asq_broker.consume(queue_name)
    first_message = next(consumer)

    # And requeue the message
    consumer.requeue([first_message])

    # Then I should be able to consume the message again immediately
    second_message = next(consumer)
    assert first_message == second_message


def test_creates_dead_letter_queue():
    # Given that I have an SQS asq_broker with dead letters turned on
    asq_broker = asq.ASQBroker(dead_letter=True)

    # When I create a queue
    asq_broker.declare_queue("test")

    # Then a dead-letter queue should be created
    dlq = asq._get_dlq_client("test")
    assert isinstance(dlq, QueueClient)


def test_consumer_returns_none_with_empty_queue(queue_name):
    asq_broker = asq.ASQBroker(dead_letter=False)
    asq_broker.declare_queue(queue_name)

    # Consume a message off an empty queue
    consumer = asq_broker.consume(queue_name)
    first_message = next(consumer)

    # As the queue is empty message should be None
    assert not first_message

    # Try another time
    second_message = next(consumer)
    assert not second_message


def test_flush_queues_returns_no_message(asq_broker, queue_name):
    # Given an actor
    @dramatiq.actor(queue_name=queue_name)
    def do_work():
        pass

    # When I send that actor messages
    for _ in range(20):
        do_work.send()

    # And flush the queue
    asq_broker.flush_all()

    # Then the consumer returns no message
    consumer = asq_broker.consume(queue_name)
    assert not next(consumer)


def test_invalid_queue_fails(queue_name):
    # Given a broker
    asq_broker = asq.ASQBroker(dead_letter=False)

    # When I attempt to consume from an undeclared queue
    # Then an exception is raised
    with pytest.raises(dramatiq.errors.QueueNotFound):
        asq_broker.validate_queue(queue_name)


def test_redeclare_queue_passes(queue_name):
    # Given a broker and a declared queue
    asq_broker = asq.ASQBroker(dead_letter=True)
    asq_broker.declare_queue(queue_name)

    # When I attempt to redeclare an existing queue
    asq_broker.declare_queue(queue_name)
    asq_broker.declare_queue(queue_name)
    asq_broker.declare_queue(queue_name)

    # Then there's only one queue created
    assert len(asq_broker.queues) == 1


def test_nacked_messages_go_to_dlq(queue_name):
    # Given an actor and a consumer, and some queued messages
    asq_broker = asq.ASQBroker(dead_letter=True)
    asq_broker.declare_queue(queue_name)

    @dramatiq.actor(queue_name=queue_name)
    def do_work():
        pass

    for i in range(20):
        do_work.send(i)
    consumer = asq_broker.consume(queue_name)

    msg = next(consumer)
    msg_content = msg._asq_message.content
    assert msg.message_id in consumer.queued_message_ids

    # When I nack a message
    consumer.nack(msg)

    # Then the message is moved to the DL queue
    assert msg.message_id not in consumer.queued_message_ids
    dlq = asq._get_dlq_client(queue_name)
    dlqd_msg = dlq.receive_message()
    assert dlqd_msg.content == msg_content


tests/test_broker.py
import json
import time

import dramatiq
import pytest
from azure.storage.queue import QueueClient

from dramatiq_azure import asq


def test_can_enqueue_and_process_messages(asq_broker, worker, queue_name):
    # Given that I have an actor that stores incoming messages in a database
    db = []

    @dramatiq.actor(queue_name=queue_name)
    def do_work(x):
        db.append(x)

    # When I send that actor a message
    do_work.send(1)

    # And wait for it to be processed
    time.sleep(5)

    # Then the db should contain that message
    assert db == [1]


def test_limits_prefetch_if_consumer_queue_is_full(asq_broker, worker, queue_name):
    # Given that I have an actor that stores incoming messages in a database
    db = []

    # Set the worker prefetch limit to 1
    worker.queue_prefetch = 1

    # Add delay to actor logic to simulate processing time
    @dramatiq.actor(queue_name=queue_name)
    def do_work(x):
        db.append(x)
        time.sleep(10)

    # When I send that actor messages, it'll only prefetch and process a single message
    do_work.send(1)
    do_work.send(2)

    # Wait for message to be processed
    time.sleep(5)

    # Then the db should contain only that message, while it sleeps
    assert len(db) == 1


def test_can_enqueue_delayed_messages(asq_broker, worker, queue_name):
    # Given that I have an actor that stores incoming messages in a database
    db = []

    @dramatiq.actor(queue_name=queue_name)
    def do_work(x):
        db.append(x)

    # When I send that actor a delayed message
    start_time = time.time()
    do_work.send_with_options(args=(1,), delay=5000)

    # And poll the database for a result each second
    for _ in range(60):
        if db:
            break

        time.sleep(1)

    # Then the db should contain that message
    assert db == [1]

    # And an appropriate amount of time should have passed
    delta = time.time() - start_time
    assert delta >= 5


def test_cant_delay_messages_for_longer_than_7_days(asq_broker, queue_name):
    # Given that I have an actor
    @dramatiq.actor(queue_name=queue_name)
    def do_work():
        pass

    # When I attempt to send that actor a message farther than 7 days into the future
    # Then I should get back a RuntimeError
    with pytest.raises(RuntimeError):
        do_work.send_with_options(delay=7 * 25 * 60 * 60 * 1000)


def test_cant_enqueue_messages_that_are_too_large(asq_broker, queue_name):
    # Given that I have an actor
    @dramatiq.actor(queue_name=queue_name)
    def do_work(s):
        pass

    # When I attempt to send that actor a message that's too large
    # Then a RuntimeError should be raised
    with pytest.raises(RuntimeError):
        do_work.send("a" * 64 * 1024)


def test_can_requeue_consumed_messages(asq_broker, queue_name):
    # Given that I have an actor
    @dramatiq.actor(queue_name=queue_name)
    def do_work():
        pass

    # When I send that actor a message
    do_work.send()

    # And consume the message off the queue
    consumer = asq_broker.consume(queue_name)
    first_message = next(consumer)

    # And requeue the message
    consumer.requeue([first_message])

    # Then I should be able to consume the message again immediately
    second_message = next(consumer)
    assert first_message == second_message


def test_creates_dead_letter_queue():
    # Given that I have an SQS asq_broker with dead letters turned on
    asq_broker = asq.ASQBroker(dead_letter=True)

    # When I create a queue
    # Then a dead-letter queue should be created
    asq_broker.declare_queue("test")

    dlq = asq._get_dlq_client("test")
    assert isinstance(dlq, QueueClient)


def test_consumer_returns_none_with_empty_queue(queue_name):
    asq_broker = asq.ASQBroker(dead_letter=False)
    asq_broker.declare_queue(queue_name)

    # Consume a message off an empty queue
    consumer = asq_broker.consume(queue_name)
    first_message = next(consumer)

    # As the queue is empty message should be None
    assert not first_message

    # Try another time
    second_message = next(consumer)
    assert not second_message


dramatiq_azure/asq.py
import logging
import os
import time
from dataclasses import dataclass
from typing import (
    Iterable,
    List,
    Optional,
)

import dramatiq
from azure.core.exceptions import (
    HttpResponseError,
    ResourceExistsError,
)
from azure.storage.queue import (
    BinaryBase64DecodePolicy,
    BinaryBase64EncodePolicy,
    QueueClient,
    QueueMessage,
)
from dramatiq.common import compute_backoff

# Set the logging level for all azure-storage-* libraries
azure_logger = logging.getLogger("azure")
azure_logger.setLevel(logging.WARNING)


logger = logging.getLogger(__name__)

#: The max number of messages that may be prefetched at a time.
MAX_PREFETCH = 32


#: The minimum time to wait between polls in second.
MIN_TIMEOUT = int(os.getenv("DRAMATIQ_ASQ_MIN_TIMEOUT", "20"))

#: to the dead-letter queue (if enabled).
MAX_RECEIVES = 3

#: Azure Storage authentication
ENDPOINT_SUFFIX = os.getenv("AZURE_ENDPOINT_SUFFIX", "core.windows.net")
CONN_STR = os.getenv("AZURE_STORAGE_CONNECTION_STR", "")
ACCOUNT_NAME = os.getenv(
    "AZURE_STORAGE_ACCOUNT_NAME",
    os.getenv("AZURE_ACCOUNT_NAME", ""),
)
USE_SSL = os.getenv("AZURE_SSL", "true").lower() == "true"
PROTOCOL = "https" if USE_SSL else "http"
ACCOUNT_URL = os.getenv(
    "AZURE_QUEUE_ACCOUNT_URL",
    f"{PROTOCOL}://{ACCOUNT_NAME}.queue.{ENDPOINT_SUFFIX}",
)


def _get_client(queue_name: str) -> QueueClient:
    if CONN_STR:
        return QueueClient.from_connection_string(
            conn_str=CONN_STR,
            queue_name=queue_name,
            message_encode_policy=BinaryBase64EncodePolicy(),
            message_decode_policy=BinaryBase64DecodePolicy(),
        )
    else:
        from azure.identity import DefaultAzureCredential

        return QueueClient(
            account_url=ACCOUNT_URL,
            queue_name=queue_name,
            credential=DefaultAzureCredential(),  # type: ignore
            message_encode_policy=BinaryBase64EncodePolicy(),
            message_decode_policy=BinaryBase64DecodePolicy(),
        )


def _get_dlq_client(queue_name: str) -> QueueClient:
    dlqueue_name = f"{queue_name}-dlq"
    return _get_client(dlqueue_name)


@dataclass
class ConsumerOptions:
    queue_name: str
    prefetch: int
    timeout: int
    dead_letter: bool = False


class _ASQMessage(dramatiq.MessageProxy):
    def __init__(
        self, asq_message: QueueMessage, message: dramatiq.Message
    ) -> None:
        super().__init__(message)
        # force type hint
        self.message_id = message.message_id
        self._message = message
        self._asq_message = asq_message

    @classmethod
    def from_queue_message(cls, _message: QueueMessage):
        dramatiq_message = dramatiq.Message.decode(_message.content)
        return cls(_message, dramatiq_message)


class ASQConsumer(dramatiq.Consumer):
    def __init__(
        self, broker: dramatiq.Broker, options: ConsumerOptions
    ) -> None:
        self.prefetch = min(options.prefetch, MAX_PREFETCH)
        self.timeout = options.timeout
        self.visibility_timeout = int(options.timeout / 1000)
        self.queue_name = options.queue_name
        self.dead_letter = options.dead_letter
        self.q_client = _get_client(options.queue_name)
        self.dlq_client = (
            _get_dlq_client(options.queue_name) if options.dead_letter else None
        )

        # local cache
        self.message_cache: List[_ASQMessage] = []
        self.queued_message_ids = set()
        self.misses = 0

    @property
    def outstanding_message_count(self):
        return len(self.queued_message_ids) + len(self.message_cache)

    def __remove_from_queue(self, message: _ASQMessage):
        self.q_client.delete_message(message._asq_message)
        if message.message_id in self.queued_message_ids:
            self.queued_message_ids.remove(message.message_id)

    def ack(self, message: _ASQMessage) -> None:
        self.__remove_from_queue(message)

    def nack(self, message: _ASQMessage) -> None:
        """
        Send to the dead-letter queue, if available.
        Dead-leter queues are meant to be managed manually.
        """
        if self.dlq_client is not None:
            self.dlq_client.send_message(message._message.encode())
        self.__remove_from_queue(message)

    def requeue(self, messages: Iterable[_ASQMessage]) -> None:
        # No batch processing
        for message in messages:
            self.q_client.send_message(message._message.encode())
            self.__remove_from_queue(message)

    def __next__(self) -> Optional[_ASQMessage]:
        while True:
            try:
                match = self.message_cache.pop(0)
                self.misses = 0
                self.queued_message_ids.add(match.message_id)
                return match
            except IndexError:
                msg_batch = []
                if self.outstanding_message_count < self.prefetch:
                    fillout = self.prefetch - self.outstanding_message_count
                    kw = {"messages_per_page": fillout}
                    if self.visibility_timeout is not None:
                        kw["visibility_timeout"] = self.visibility_timeout
                    pager = self.q_client.receive_messages(**kw)
                    try:
                        msg_batch = [item for item in next(pager.by_page())]
                        self.message_cache = [
                            _ASQMessage.from_queue_message(_msg)
                            for _msg in msg_batch
                        ]
                    except StopIteration:
                        self.message_cache = []

                if not msg_batch:
                    self.misses, backoff_ms = compute_backoff(
                        self.misses, max_backoff=self.timeout
                    )
                    time.sleep(backoff_ms / 1000)
                    return None


class ASQBroker(dramatiq.Broker):
    """A Dramatiq_ broker that can be used with `Azure Storage Queues`_
    This backend has a number of limitations compared to the built-in
    Redis and RMQ backends:
      * messages can be at most 64KiB large,
    The backend uses the `Python Azure SDK`_ (v12).
    Parameters:
      middleware: The set of middleware that apply to this broker.
      dead_letter: Whether to add a dead-letter queue. Defaults to false.
    .. _Dramatiq: https://dramatiq.io
    .. _Azure Storage Queues: https://docs.microsoft.com/en-us/azure/storage/queues/
    .. _Python Azure SDK: https://github.com/Azure/azure-sdk-for-python
    """

    def __init__(
        self,
        *,
        dead_letter: bool = False,
        middleware=None,
    ) -> None:
        super().__init__(middleware=middleware)
        self.queues: set = set()
        self.dead_letter = dead_letter

    @property
    def consumer_class(self):
        """Allows overriding the default consumer"""
        return ASQConsumer

    def validate_queue(self, queue_name: str):
        if queue_name not in self.queues:
            raise dramatiq.errors.QueueNotFound(queue_name)

    def consume(
        self, queue_name: str, prefetch: int = 1, timeout: int = 5000
    ) -> dramatiq.Consumer:
        self.validate_queue(queue_name)
        options = ConsumerOptions(
            queue_name=queue_name,
            prefetch=prefetch,
            timeout=timeout,
            dead_letter=self.dead_letter,
        )
        return self.consumer_class(self, options)

    def declare_queue(self, queue_name: str) -> None:
        if queue_name not in self.queues:
            self.emit_before("declare_queue", queue_name)
            try:
                q_client = _get_client(queue_name)
                q_client.create_queue()
            except ResourceExistsError:
                logger.warning(f"Queue already exists: {queue_name}")
            if self.dead_letter:
                try:
                    dlq_client = _get_dlq_client(queue_name)
                    dlq_client.create_queue()
                except ResourceExistsError:
                    logger.warning(f"DL Queue already exists: {queue_name}")
            self.queues.add(queue_name)
            self.emit_after("declare_queue", queue_name)

    def enqueue(
        self, message: dramatiq.Message, *, delay: Optional[int] = None
    ) -> dramatiq.Message:
        queue_name = message.queue_name
        self.validate_queue(queue_name)

        delay_sec = int(delay / 1000) if delay else 0

        logger.debug(
            f"Enqueueing message {message.message_id} on queue {queue_name}."
        )
        self.emit_before("enqueue", message, delay)
        q_client = _get_client(queue_name)
        try:
            q_client.send_message(
                message.encode(), visibility_timeout=delay_sec
            )
            self.emit_after("enqueue", message, delay)
            return message
        except HttpResponseError as e:
            raise RuntimeError(str(e))

    def flush(self, queue_name: str):
        self.validate_queue(queue_name)
        q_client = _get_client(queue_name)
        q_client.clear_messages()

    def flush_all(self):
        for queue_name in self.queues:
            self.flush(queue_name)

    def get_declared_queues(self) -> Iterable[str]:
        return self.queues

    def get_declared_delay_queues(self) -> Iterable[str]:
        return set()
